### PR TITLE
[RISCV] full support for riscv_rvv_vector_bits attribute

### PR DESCRIPTION
**Pull request by vbe-sc**
Add support for using attribute((rvv_vector_bits(N))), when N < 8.
It allows using all fixed length vector mask types regardless VLEN value.

### CODE REVIEW

**Comment by vbe-sc** [comment]
@topperc, could you please have a look at this patch

File: clang/lib/AST/ASTContext.cpp [hunk 1]
```diff
@@ -10014,14 +10023,33 @@ bool ASTContext::areCompatibleRVVTypes(QualType FirstType,
...
+        }
+        if (VT->getVectorKind() == VectorKind::RVVFixedLengthMask_4) {
+          BuiltinVectorTypeInfo Info = getBuiltinVectorTypeInfo(BT);
+          return FirstType->isRVVVLSBuiltinType() &&
+                 Info.ElementType == BoolTy &&
+                 getTypeSize(SecondType) == ((getRVVTypeSize(*this, BT)) * 2);
         }
         if (VT->getVectorKind() == VectorKind::RVVFixedLengthData ||
-            VT->getVectorKind() == VectorKind::Generic)
+            VT->getVectorKind() == VectorKind::Generic) {
```
Comment (topperc) [inline clang/lib/AST/ASTContext.cpp:10047]: Curly braces aren't supposed to be used on single line bodies per coding standards https://llvm.org/docs/CodingStandards.html#don-t-use-braces-on-simple-single-statement-bodies-of-if-else-loop-statements

File: clang/test/CodeGen/RISCV/attr-riscv-rvv-vector-bits-less-8-cast.c [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+//
+// CHECK-128-LABEL: @from_vbool32_t(
+// CHECK-128-NEXT:  entry:
+// CHECK-128-NEXT:    [[SAVED_VALUE:%.*]] = alloca <vscale x 2 x i1>, align 1
+// CHECK-128-NEXT:    [[RETVAL_COERCE:%.*]] = alloca <vscale x 4 x i1>, align 1
+// CHECK-128-NEXT:    store <vscale x 2 x i1> [[TYPE:%.*]], ptr [[SAVED_VALUE]], align 1, !tbaa [[TBAA6:![0-9]+]]
+// CHECK-128-NEXT:    [[TMP0:%.*]] = load <1 x i8>, ptr [[SAVED_VALUE]], align 1, !tbaa [[TBAA10:![0-9]+]]
+// CHECK-128-NEXT:    store <1 x i8> [[TMP0]], ptr [[RETVAL_COERCE]], align 1
+// CHECK-128-NEXT:    [[TMP1:%.*]] = load <vscale x 4 x i1>, ptr [[RETVAL_COERCE]], align 1
+// CHECK-128-NEXT:    ret <vscale x 4 x i1> [[TMP1]]
```
Comment (topperc) [inline clang/test/CodeGen/RISCV/attr-riscv-rvv-vector-bits-less-8-cast.c:31]: Why is the return type vscale x 4 x i1?

File: clang/lib/AST/ASTContext.cpp (hunk 1 above)
Comment (vbe-sc) [reply clang/lib/AST/ASTContext.cpp]: Fix

File: clang/test/CodeGen/RISCV/attr-riscv-rvv-vector-bits-less-8-cast.c (hunk 2 above)
Comment (vbe-sc) [reply clang/test/CodeGen/RISCV/attr-riscv-rvv-vector-bits-less-8-cast.c:48]: You are right. There was a bug, I didn't change the policy for `RISCVABIInfo::coerceVLSVector`.

File: clang/test/CodeGen/RISCV/attr-rvv-vector-bits-types.c [hunk 3]
```diff
@@ -572,1483 +1121,44 @@ void f() {
 //===----------------------------------------------------------------------===//
 // Structs and unions
 //===----------------------------------------------------------------------===//
-// CHECK-64:      %struct.struct_int8m1 = type { <8 x i8> }
```
Comment (topperc) [inline clang/test/CodeGen/RISCV/attr-rvv-vector-bits-types.c:575]: Using the script erased a large portion of the CHECKs in this test. The script only knows how to check function bodies. It doesn't know about global variables or types.

File: clang/test/CodeGenCXX/riscv-mangle-rvv-fixed-vectors.cpp [hunk 4]
```diff
@@ -176,477 +177,3110 @@ typedef vbool1_t fixed_bool1_t __attribute__((riscv_rvv_vector_bits(__riscv_v_fi
...
-#if __riscv_v_fixed_vlen >= 256
 typedef vbool32_t fixed_bool32_t __attribute__((riscv_rvv_vector_bits(__riscv_v_fixed_vlen/32)));
-#endif
-#if __riscv_v_fixed_vlen >= 512
 typedef vbool64_t fixed_bool64_t __attribute__((riscv_rvv_vector_bits(__riscv_v_fixed_vlen/64)));
-#endif
 
 template <typename T> struct S {};
 
-// CHECK-64: _Z2f11SI9__RVV_VLSIu14__rvv_int8m1_tLj64EEE
```
Comment (topperc) [inline clang/test/CodeGenCXX/riscv-mangle-rvv-fixed-vectors.cpp:191]: This test was only testing the mangling names not the body. That's why it didn't use the update script.

File: clang/test/CodeGen/RISCV/attr-rvv-vector-bits-types.c (hunk 3 above)
Comment (vbe-sc) [reply clang/test/CodeGen/RISCV/attr-rvv-vector-bits-types.c]: Done

File: clang/test/CodeGenCXX/riscv-mangle-rvv-fixed-vectors.cpp (hunk 4 above)
Comment (vbe-sc) [reply clang/test/CodeGenCXX/riscv-mangle-rvv-fixed-vectors.cpp]: Many thanks for your comment: it allowed me to understand this test and find an improvement in `clang/./lib/AST/ItaniumMangle.cpp` for my patch. 
Fixed

**Comment by vbe-sc** [comment]
@topperc, ping

File: clang/test/CodeGenCXX/riscv-mangle-rvv-fixed-vectors.cpp [hunk 5]
```diff
@@ -1,7 +1,7 @@
 // RUN: %clang_cc1 -triple riscv64-none-linux-gnu %s -emit-llvm -o - \
 // RUN:  -target-feature +f -target-feature +d -target-feature +zfh \
 // RUN:  -target-feature +zve64d -target-feature +zvfh -mvscale-min=1 \
-// RUN:   -mvscale-max=1 | FileCheck %s --check-prefix=CHECK-64
+// RUN:   -mvscale-max=1 | FileCheck %s --check-prefix=CHECK-64
```
Comment (topperc) [inline clang/test/CodeGenCXX/riscv-mangle-rvv-fixed-vectors.cpp:4]: Stray change?

Comment (vbe-sc) [reply clang/test/CodeGenCXX/riscv-mangle-rvv-fixed-vectors.cpp]: Done

**Review by topperc — APPROVED** [review]
LGTM

**Pull request merged**